### ISSUE TITLE
Type narrowing is too aggressive

### ISSUE THREAD

**Issue by hurryabit** [microsoft/pyright#9193]
Pyright does not report any type errors in the following code:

```python
class A:
    pass

class B:
    pass

class C(A):
    def only_on_c(self) -> None:
        pass

class D(A, B):
    pass

def boom(b_or_c: B | C) -> None:
    if isinstance(b_or_c, A):
        # reveal_type(b_or_c)
        b_or_c.only_on_c()

boom(D())
```

This code fails at runtime though since class `D` does not have a method named `only_on_c`. Uncommenting the `reveal_type` surfaces that pyright thinks `b_or_c` has type `C` at this point. This type is too narrow since it discards the option `A & B`, where `&` is type intersection.

I'm using pyright 1.1.384 installed via `pip`. The version running in my VS Code shows the same behaviour.

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

From a type theory standpoint, you're correct that the narrowed type in this case is "too narrow". To be completely correct, it would need to be `(A & B) | C` (where `&`) represents an intersection type. This would catch the crash in your code sample, eliminating the false negative. However, it would create many false positives in well-written Python code. Pyright make pragmatic tradeoffs in this case and opt to eliminate frequent false positives at the expense of some rare false negatives.

You will find the same behavior in mypy and other Python type checkers.
